[Feature] Add quantum information tools

inafergra opened the pull request:
This PR follows the work of this [old qadence PR](https://github.com/pasqal-io/qadence/pull/305). It adds a subpackage `qinfo_tools` containing different quantum information-related functionalities. The three files of the subpackage are:
- `qadence_libs/qinfo_tools/qfi.py` contains code related to the calculation of the Quantum Fisher Information (QFI) matrix. The function `get_quantum_fisher()` calculates the QFI matrix of a `QuantumCircuit` in an exact way, while the `get_quantum_fisher_spsa()` can be used to iteratively construct an approximation of the QFI matrix through the [SPSA](https://www.jhuapl.edu/spsa/) algorithm. 
 - `qadence_libs/qinfo_tools/spsa.py`contains the code to calculate the SPSA gradients. Only the second order gradient is used in the QNG-SPSA optimizer but the first order gradient is also implemented for completeness. 
 - The main contribution is the `qadence_libs/qinfo_tools/qng.py` file implementing the Quantum Natural Gradient (QNG) and the QNG-SPSA optimizers. Both optimizers use the QFI matrix in the parameter update rule to better navigate the gradient descent. The QNG uses the exact QFI matrix, which takes $\mathcal{O}(n^2)$ circuit evaluations per iteration ($n$ being the number of variational params in the circuit), making the QNG optimizer very slow. The QNG-SPSA uses the SPSA approximation of the QFI matrix (https://arxiv.org/pdf/2103.09232.pdf) resulting a much more efficient optimizer: it only takes 6 circuit evaluations per iteration to calculate the approximation of the QFI matrix (constant overhead $\mathcal{O}(1)$ ).

A new section in the docs has also been created for quantum information related topics, with a file explaining how to use the QNG and QNG-SPSA optimizers for an easy QCL problem. Since there was yet no files in the documentation, this PR also adds a `docs/docsutils.py` file with a `fig_to_html()` function to print images in the docs (same as in the Qadence repo) and a `docs/environment.yml` for the python environment (also same as in Qadence).

Roland-djee commented:
Hey @inafergra thank you for this. Is this ready for review ? In case, you should un-draft it.

inafergra commented:
> Hey @inafergra thank you for this. Is this ready for review ? In case, you should un-draft it.

It is ready for review now

gvelikova commented:
Thanks Ignacio! I think you should add the qinfo_tools imports to the __init__.py of the package in the same way as the  constructors. I am testing it with some PDEs and will try to add a proper review soon.

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+
+    or batch of matrices.
+
+    Code from https://github.com/pytorch/pytorch/issues/25481#issuecomment-1032789228
```
Is this code stolen form that issue ?

inafergra replied:
Pretty much. Is this okay? There's not really any functionality in torch to compute the square of a matrix so this was what I found. It was posted by a torch developer in the issue.

inafergra replied:
I did do some checks and it worked as intended. I could also add a test in `tests` for the function to be sure

inafergra replied:
I added a test for this.

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    """Sets the variational parameter values of the circuit."""
+    if vparams_values is not None:
+        blocks = primitive_blocks(circuit.block)
+        iter_index = iter(range(len(blocks)))
```
Can't you use `enumerate` ?

inafergra replied:
The issue with using `enumerate` is that some parameters `p` in the loop are not trainable, while `vparams_values` only holds the values of the trainable ones. This means that the index would not match the corresponding value in `p.value` as with every non-trainable parameter the index would jump one ahead. Using an iterator solves this problem.

inafergra replied:
The QFI functions now get a dictionary as input.

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+
+def get_quantum_fisher(
+    circuit: QuantumCircuit,
+    vparams_values: Iterable[float | Tensor] | None = None,
```
I think in Qadence, parameter values are dicts.

inafergra replied:
Right now, this function is only used in the parameter update rule of the QNG optimizer and is meant to take the values of the trainable parameters as given by torch's `Optimizer` class, which are not given as dicts. Would it be better to change the name of the argument for consistent nomenclature?

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+                    p.value = float(vparams_values[next(iter_index)])  # type: ignore
+
+
+def _get_fm_dict(circuit: QuantumCircuit) -> dict:
```
@dominikandreasseitz We have something built-in Qadence for that I think ?

dominikandreasseitz replied:
actually not directly

gvelikova replied:
There is the reset_vparams method in the QuantumModel

inafergra replied:
@dominikandreasseitz what do you think?

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    overlap_method: OverlapMethod = OverlapMethod.EXACT,
+    diff_mode: DiffMode = DiffMode.AD,  # type: ignore
+) -> Tensor:
+    """Returns the exact Quantum Fisher Information (QFI) matrix of the quantum circuit.
```
Docstring formatting.

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+def get_quantum_fisher(
+    circuit: QuantumCircuit,
+    vparams_values: Iterable[float | Tensor] | None = None,
+    fm_dict: dict[str, Tensor] = {},
```
```suggestion
    fm_dict: dict[str, Tensor] = dict(),
```

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    """
+
+    # Get feature map dictionary (required to run Overlap().forward())
+    if fm_dict == {}:
```
Are you checking for emptiness ? In case, I'll do:

```suggestion
    if not fm_dict:
```

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+
+    # Get Overlap() model
+    ovrlp_model = Overlap(
+        circuit,
```
Don't you want to dagger your circuit ? Might be saying something stupid though.

inafergra replied:
The `Overlap` class already takes care of doing the dagger (see [here](https://github.com/pasqal-io/qadence/blob/a92780f75da81bce26f83a99fd5b46542531cdff/qadence/overlap.py#L178))

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    # Calculate the Hessian
+    fid_hess = hessian(ovrlp, vparams)
+
+    return -2 * fid_hess
```
```suggestion
    return -2 * hessian(ovrlp, vparams)
```

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    iteration: int,
+    vparams_values: Iterable[float | Tensor] | None = None,
+    fm_dict: dict[str, Tensor] = {},
+    previous_qfi_estimator: Tensor = None,
```
```suggestion
    previous_qfi_estimator: Tensor | None = None,
```

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    overlap_method: OverlapMethod = OverlapMethod.EXACT,
+    diff_mode: DiffMode = DiffMode.AD,  # type: ignore
+) -> Tensor:
+    """Function to calculate the Quantum Fisher Information (QFI) matrix with the.
```
Docstring.

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    circuit: QuantumCircuit,
+    iteration: int,
+    vparams_values: Iterable[float | Tensor] | None = None,
+    fm_dict: dict[str, Tensor] = {},
```
```suggestion
    fm_dict: dict[str, Tensor] = dict(),
```

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+        diff_mode (DiffMode, optional): Defaults to DiffMode.ad.
+    """
+    # Get feature map dictionary (required to run Overlap().forward())
+    if fm_dict == {}:
```
See other check for emptiness.

Roland-djee commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,166 @@
+    qfi_mat_positive_sd = qfi_mat_positive_sd + beta * torch.eye(ovrlp_model.num_vparams)
+    qfi_mat_positive_sd = qfi_mat_positive_sd / (1 + beta)  # regularization
+
+    return qfi_mat_estimator, qfi_mat_positive_sd
```
Is this return type a `Tensor` ?

inafergra replied:
Fixed

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,132 @@
+    def __init__(
+        self,
+        params: tuple | torch.Tensor,
+        circuit: QuantumCircuit = required,
```
Is this default value intrinsic to `Optimizer` ?

inafergra replied:
[Yes](https://github.com/pytorch/pytorch/blob/main/torch/optim/optimizer.py#L49C1-L49C6)

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,132 @@
+    def __setstate__(self, state):  # type: ignore
+        super().__setstate__(state)
+
+    def step(self, closure: Callable | None = None) -> torch.Tensor:
```
```suggestion
    def step(self, closure: callable | None = None) -> torch.Tensor:
```

inafergra replied:
Is this right? The `Callable` type from `typing.Callable` is with capital C (https://docs.python.org/3/library/typing.html#annotating-callable-objects)

Roland-djee replied:
True enough but since you do:

```python
from __future__ import annotations
```
I thought there was a `callable` type. Maybe I'm wrong here.

Roland-djee replied:
```suggestion
    def step(self, closure: Callable | None = None) -> Tensor:
```

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,132 @@
+        return loss
+
+
+class QNG_SPSA(Optimizer):
```
Couldn't you make that one a derived class from `QNG` ?

inafergra replied:
Not sure about this. There is not much to inherit from `QNG` and I believe it's cleaner to separate both. Also, this is how it is usually implemented in Pytorch, just directly inheriting from the `Optimizer` class (see [Adam](https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam) and [AdamW](https://pytorch.org/docs/stable/_modules/torch/optim/adamw.html#AdamW) implementations in Pytorch)

Roland-djee replied:
I kinda disagree: the initialiser is almost identical, `__setstate__` is common, only `step` differs but there are commonalities. I think you can design an abstract class (possibly a dataclass) with an extra argument to select either the standard `QNG` or the `SPSA` variant.

Roland-djee commented on qadence_libs/qinfo_tools/utils.py:
```diff
@@ -0,0 +1,40 @@
+from torch.autograd import grad
+
+
+def hessian(output: Tensor, inputs: list) -> Tensor:
```
Could you specify the list type ?

Roland-djee commented on qadence_libs/qinfo_tools/utils.py:
```diff
@@ -0,0 +1,40 @@
+
+    n_params = len(inputs)
+    hess = torch.empty((n_params, n_params))
+    for i in range(n_params):
```
Maybe worth adding a `jacobian_var = jacobian[i]` since it is used twice ?

Roland-djee commented on tests/qinfo_tools/test_qfi.py:
```diff
@@ -0,0 +1,77 @@
+random.seed(SEED)
+
+
+def create_hea_circuit(n_qubits: int, layers: int) -> QuantumCircuit:
```
Maybe you can use [this](https://github.com/pasqal-io/qadence/blob/a92780f75da81bce26f83a99fd5b46542531cdff/qadence/constructors/ansatze.py#L17) ?

inafergra replied:
I'm using it inside the function, but `hea()` outputs an `AbstractBlock` with no FM and basically what I do inside the function is add a FM and wrap everything in a `QuantumCircuit` since that's what I need

Roland-djee commented on tests/qinfo_tools/test_qng.py:
```diff
@@ -0,0 +1,76 @@
+random.seed(SEED)
+
+
+def quadratic_dataset(samples: int) -> tuple[Tensor, Tensor]:
```
If they are reused multiple times, I'd add them as fixtures to  `conftest.py`

Roland-djee commented on tests/qinfo_tools/test_qng.py:
```diff
@@ -0,0 +1,76 @@
+
+@pytest.mark.parametrize("dataset", DATASETS)
+@pytest.mark.parametrize("optim_config", OPTIMIZERS_CONFIG)
+@pytest.mark.parametrize("n_qubits", [2])
```
Is it worth parametrising for a single value ?

Roland-djee reviewed (commented):
Thanks a million @inafergra. Comments from my side. I'd also like to see more tests for the functionality you use throughout. Atm, you test your top-level datastructures but you assume some of the functionality they use is correct. I would make sure this is also the case.

dominikandreasseitz reviewed (approved):
thanks @inafergra, great work.  did you get to compare runtime and memory to adam?

inafergra commented:
> Thanks a million @inafergra. Comments from my side. I'd also like to see more tests for the functionality you use throughout. Atm, you test your top-level datastructures but you assume some of the functionality they use is correct. I would make sure this is also the case.

Thanks a lot for your comments, Roland. I will try to find some time this week to go through your comments and add more tests.

inafergra commented:
> thanks @inafergra, great work. did you get to compare runtime and memory to adam?

Thank you, Dominik. I haven't had the time yet but I'll try to find some time this week to do it.

inafergra commented on docs/qinfo_tools/qng.md:
```diff
@@ -0,0 +1,174 @@
+model_adam = qd.QNN(circuit_adam, [observable])
+
+# QNG
+circuit_qng = qd.QuantumCircuit(n_qubits, feature_map, ansatz)
```
No! I thought I had to instantiate different circuits to keep the model parameters independent, but it works using the same circuit

inafergra replied:
I changed it so all circuits are the same

Roland-djee commented:
Hey @inafergra will you have time to work on this per chance ? I am keen on not letting it linger for too long.

inafergra commented:
> Hey @inafergra will you have time to work on this per chance ? I am keen on not letting it linger for too long.

Hey Roland, sorry, client project got in the way. If there's a preference for merging it soon I'll try to find some time this week to make the last changes and request a final review.

Roland-djee commented:
> > Hey @inafergra will you have time to work on this per chance ? I am keen on not letting it linger for too long.
> 
> Hey Roland, sorry, client project got in the way. If there's a preference for merging it soon I'll try to find some time this week to make the last changes and request a final review.

No rush. If you could that soon, that'd be great

dominikandreasseitz commented:
@inafergra @gvelikova @Roland-djee,  lets try to wrap this up: @inafergra , can you list some the remaining todos? (if there are any). This is a huge amount of work, if some usecases for example dont converge yet, i would mark it as an experimental feature and list a number of proven working usecases. we can iteratively improve this optimizer once we get people to use it

inafergra commented:
> @inafergra @gvelikova @Roland-djee, lets try to wrap this up: @inafergra , can you list some the remaining todos? (if there are any). This is a huge amount of work, if some usecases for example dont converge yet, i would mark it as an experimental feature and list a number of proven working usecases. we can iteratively improve this optimizer once we get people to use it

Apart from a couple of minor comments that need to be addressed (but should be fast), the two main TODOS are still:

- Adding more tests for the low level functionalities as suggested by Roland
- Benchmark memory consumption of QNG and QNG-SPSA compared to ADAM

gvelikova reviewed (approved):
Other than the issue that I mentioned to @inaferga on me going over few more PDE examples for sanity checking of correctness, I think it is a pretty good work that can be merged.

inafergra commented:
@Roland-djee I've addressed all the comments and requested your review once again. We had a meeting with @dominikandreasseitz and @gvelikova today and they are okay with merging it in its current state. There's still many ways the PR can be improved but we can iterate over it once people start using and we receive their feedback.

inafergra commented:
This are the time and memory benchmarks for circuits with constant depth
![time](https://github.com/pasqal-io/qadence-libs/assets/51716758/5eb07806-ac6b-4036-9335-b0a9e0be93eb)
![memory](https://github.com/pasqal-io/qadence-libs/assets/51716758/5e34ebd3-d621-46fc-b440-592ef31f936e)

The exact QNG is very costly both in time and memory and we should make it clear that it is not meant to be used with medium to large circuits. I've added a warning in the docstring about this. The SPSA approximation is much more efficient while still retaining a good performance

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,182 @@
+from qadence_libs.qinfo_tools.qfi import get_quantum_fisher, get_quantum_fisher_spsa
+
+
+class QNG(Optimizer):
```
```suggestion
class QuantumNaturalGradient(Optimizer):
```

People could alias it with `QNG` if they wish to.

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,182 @@
+                Defaults to 10e-3.
+        """
+
+        if not 0.0 <= lr:
```
Isn't it simpler to do:

```suggestion
        if 0.0 > lr:
```

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,182 @@
+        if not 0.0 <= beta:
+            raise ValueError(f"Invalid beta value: {beta}")
+        if not isinstance(circuit, QuantumCircuit):
+            raise ValueError("The circuit should be an instance of qadence.QuantumCircuit")
```
```suggestion
            raise ValueError(f"The circuit should be an instance of {type(QuantumCircuit)}")
```

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,182 @@
+            raise ValueError("The circuit should be an instance of qadence.QuantumCircuit")
+
+        defaults = dict(circuit=circuit, lr=lr, beta=beta)
+        super(QNG, self).__init__(params, defaults)
```
Do you still need to provide arguments to `super` ?

inafergra replied:
Yes, this is needed, see for example the [ADAM optimizer ](https://github.com/pytorch/pytorch/blob/main/torch/optim/adam.py#L72C1-L72C43). I think it is required mostly to load the state_dict

Roland-djee commented on qadence_libs/qinfo_tools/spsa.py:
```diff
@@ -0,0 +1,116 @@
+    Args:
+        size (int): Size of the vector
+    """
+    return torch.Tensor(np.random.choice([-1, 1], size=(size, 1)))
```
```suggestion
    return torch.tensor(np.random.choice([-1, 1], size=(size, 1)))
```

Roland-djee replied:
Actually, can't you use [torch.rand](https://pytorch.org/docs/stable/generated/torch.rand.html) ?

Roland-djee commented on qadence_libs/qinfo_tools/spsa.py:
```diff
@@ -0,0 +1,116 @@
+    shifted_vparams_dict = {k: (v + s) for (k, v), s in zip(vparams_dict.items(), shift)}
+
+    ovrlp_shifted = model(
+        bra_param_values=fm_dict | vparams_dict,
```
What does the `|` do here ?

inafergra replied:
It [merges](https://peps.python.org/pep-0584/) the two dictionaries

Roland-djee reviewed (commented):
Thanks @inafergra for that impressive contribution. I think we should discuss together with @dominikandreasseitz and @gvelikova so that we come up with a design for a `QuantumNaturalGradient` class. My main concern is the fragility of the vparams replacement. We should also see if that still lies here or if that's best placed elsewhere (`QuantumModel` for instance).

inafergra commented on qadence_libs/qinfo_tools/qfi.py:
```diff
@@ -0,0 +1,155 @@
+    circuit: QuantumCircuit,
+    vparams_values: Iterable[float | Tensor] | None = None,
+    fm_dict: dict[str, Tensor] = dict(),
+    backend: BackendName = BackendName.PYQTORCH,  # type: ignore
```
Not sure why but Pylance was not able to identify the type in my vscode. I'll delete it since it should work fine

inafergra commented:
> Thanks @inafergra for that impressive contribution. I think we should discuss together with @dominikandreasseitz and @gvelikova so that we come up with a design for a `QuantumNaturalGradient` class. My main concern is the fragility of the vparams replacement. We should also see if that still lies here or if that's best placed elsewhere (`QuantumModel` for instance).

Thanks again for the comments Roland. I understand your point with `vparams_values`, I agree it may not be a robust approach in general. The only reason behind it being an iterable is that torch's optimizers require the parameters to be passed as an iterable. Right now this approach relies on the user passing the variational parameters to the optimizer in the "correct" order, which will be the case if they just use the `parameters()` method in the `QuantumCircuit` class. 

Maybe we could have different a behaviour in the `get_quantum_information()` functions depending on whether the `vparams_values` is an `iterable` or a `dict`? This way, those functions can be used in general (in a robust way) to compute the QFI of a circuit (outside of the `QNG` optimizer) by passing the a `dict` of parameters. And if an `iterable` is passed then we assume the ordering given by the `parameters()` method in `QuantumCircuit`, which is the required behaviour in the QNG optimizer. I would add a test to check that the indexing orderings match and make sure it's visible in the documentation. If this proposal does not work, I'd need to rethink if there is a workaround to pass a 'dict' to torch's optimizers, but at the moment I'm not sure how can this be done.

dominikandreasseitz commented:
@inafergra https://github.com/pasqal-io/qadence/blob/main/qadence/engines/torch/differentiable_expectation.py#L33 here we do the param_dict split i was talking about

jpmoutinho commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,182 @@
+                metric_tensor = metric_tensor + group["beta"] * torch.eye(len(grad_vec))
+
+                # Get transformed gradient vector
+                metric_tensor_inv = torch.linalg.inv(metric_tensor)
+                transf_grad = torch.matmul(metric_tensor_inv, grad_vec)
```
So I came to check this after your presentation, it might be worth changing this to solve a linear system instead. Since you seem to be doing $x=A^{-1}b$ for `x=transf_grad`, typically the numerics of solving $Ax=b$ are much more efficient than those of computing $A^{-1}$.

https://www.johndcook.com/blog/2010/01/19/dont-invert-that-matrix/

jpmoutinho replied:
Same thing for the usage of `torch.linalg.pinv()`

https://pytorch.org/docs/stable/generated/torch.linalg.pinv.html
https://pytorch.org/docs/stable/generated/torch.linalg.lstsq.html#torch.linalg.lstsq

inafergra replied:
That's a great suggestion, thanks!

inafergra replied:
This has been implemented.

Roland-djee commented:
Thanks @inafergra. Is this reviewable ? Another quick question: would it make sense to move SPSA to PyQ as another differentiation method ?

dominikandreasseitz commented:
> Thanks @inafergra. Is this reviewable ? Another quick question: would it make sense to move SPSA to PyQ as another differentiation method ?

@Roland-djee  we can add spsa in pyq but the implementation in this MR is highly geared towards ignacios usecase, it for example works only for overlap models which do not exist atm in pyq

inafergra commented:
@Roland-djee I still need to add some tests related to the indexing order of the vparams. I'll mark this PR as a draft until it's done. Regarding the SPSA approximation, I agree with Dominik, it can make sense to place it in PyQ in the future but that will require generalizing the module quite a bit.

inafergra commented:
The PR is ready for review now. I have tried to implement the changes as discussed in the comments and in our meetings, mainly:
- The QFI functions now take a dictionary with the variational parameters names and tensors, to avoid dependence on the ordering of the parameters (before they were passed as a Sequence).
- However, inside the QNG optimizer the parameters still need to be passed as a Sequence (this is the expected behaviour of the `torch.optimizer ` base class). I assemble a dictionary inside the QNG optimizer to pass to the QFI functions and I make sure the ordering is correct in the tests. 
- In the parameter update we now solve the least square problems instead of computing the inverse, as Joao pointed out this is way faster and more stable. 
- I added a custom type for the different QFI approximations
- Added more tests, particularly for the symmetric matrix square root

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,181 @@
+            beta=beta,
+            epsilon=epsilon,
+        )
+        super(QuantumNaturalGradient, self).__init__(params, defaults)
```
```suggestion
        super().__init__(params, defaults)
```

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,181 @@
+                )
+
+                with torch.no_grad():
+                    # Finite shift the metric tensor to avoid problems when inverting
```
What problems ?

inafergra replied:
Numerical stability issues when inverting the matrix. We are no longer inverting the metric tensor (we solve the least square problems instead) but the shift is still needed for stability. I will change the comment accordingly.

Roland-djee commented on qadence_libs/qinfo_tools/qng.py:
```diff
@@ -0,0 +1,181 @@
+
+                    # Update parameters
+                    for i, p in enumerate(vparams_values):
+                        if p.grad is None:
```
Could this happen if `p` is a variational parameter ?

inafergra replied:
True, we already check that the elements in `vparams_values` are all trainable. I'll delete it

Roland-djee commented on qadence_libs/qinfo_tools/spsa.py:
```diff
@@ -0,0 +1,122 @@
+
+def _create_random_direction(size: int) -> Tensor:
+    """
+    Creates a torch Tensor with elements randomly drawn from {-1,+1}.
```
```suggestion
    Creates a torch Tensor with `size` elements randomly drawn from {-1,+1}.
```

Roland-djee commented on tests/qinfo_tools/test_spsa.py:
```diff
@@ -0,0 +1,49 @@
+random.seed(SEED)
+
+
+@pytest.mark.parametrize("shift", [0.0, 0.1])
```
Does it make sense to extend the range of values here ?

inafergra commented:
Thank you for the comments @Roland-djee, all addressed. Ready to merge

Roland-djee reviewed (approved):
Thanks @inafergra looking very nice.

inafergra commented:
Thanks @Roland-djee, will merge now!
